src/extension_traits/as_out.rs

```diff
@@ -0,0 +1,155 @@
+use_prelude!();
+
+use ::core::mem::ManuallyDrop;
+
+#[cfg(doc)]
+use crate::extension_traits::ManuallyDropMut;
+
+use private::Is;
+mod private {
+    pub trait Is { type Eq : ?Sized; }
+    impl<T : ?Sized> Is for T { type Eq = T; }
+}
+
+/// Helper / extension trait to convert a `&mut _` into a `&out T` by calling
+/// `.as_out::<T>()` on it.
+///
+/// By autoref, this means that you can even just extract a `&out T` reference
+/// out of a `mut` element simply by calling `.as_out::<T>()` on it.
+///
+/// There is, however, one restriction: to be able to call `.as_out::<_>()` on
+/// something, it needs to be either `Copy`, or a value wrapped in a
+/// [`MaybeUninit`] / a [`ManuallyDrop`].
+///
+///   - (or a slice of such, in which case it yields a
+///     [fat `&out` reference][`OutSlice`])
+///
+/// This is by design. Indeed, [`Out`] references do not call the destructor
+/// of the overwritten element (since it may not be initialized).
+/// This could cause memory leaks when there is an initialized element with
+/// [drop glue][`core::mem::needs_drop`].
+///
+/// To solve this limitation, one must explicitly call
+/// [`.manually_drop_mut()`][`ManuallyDropMut::manually_drop_mut`]
+/// to automagically transmute the `&mut _` reference into a
+/// `&mut ManuallyDrop<_>`.
+///
+/// # Examples
+///
+/// ```rust
+/// use ::uninit::prelude::*;
+///
+/// let mut x = 0;
+/// x.as_out::<u32>().write(42);
+///
+/// let mut y = ::core::mem::MaybeUninit::uninit();
+/// y.as_out::<u32>().write(42);
+/// let y = unsafe { y.assume_init() };
+///
+/// assert_eq!(x, y);
+/// ```
+pub
+trait AsOut<Pointee_ : ?Sized> {
+    type Out;
+
+    fn as_out<Pointee : ?Sized + Is<Eq=Pointee_>> (self: Self)
+      -> Self::Out
+    ;
+}
```

Comment:
To avoid coherence rules between `T : Copy` and `ManuallyDrop<T>`, the type `Pointee` here is a generic parameter. But then at usage site type inference struggles a lot, so it is nice to be able to use turbofish to help it. Except that the generic parameter is on the trait, so `.as_out::<>()` would not work. Hence the `<Pointee : Is<Eq=Pointee_>>` hack.

Reply:
This should be part of the documentation then, otherwise it is very hard to read. Espcially for new Rustecean which are less familiar with traits and bounds already.